fw/http_frame.c

```diff
@@ -2123,6 +2126,11 @@ tfw_h2_make_frames(struct sock *sk, struct sk_buff *skb, TfwH2Ctx *ctx,
 	TfwFrameHdr frame_hdr = {};
 	unsigned long *len = (type == HTTP2_DATA ?
 		&stream->xmit.b_len : &stream->xmit.h_len);
+	TfwFrameType otype = type;
+	if (type == HTTP2_HEADERS && stream->xmit.b_len == 0 && stream->xmit.h_len == 0) {
```

Comment:
Instead of this part of code you should implement `tfw_h2_make_trailer_frames` same as it done for header and data frames. (see tfw_h2_make_headers_frames` and `tfw_h2_make_data_frames`)

Reply:
No, `tfw_h2_make_headers_frames` and `tfw_h2_make_data_frames` do nothing but provide `type` to `tfw_h2_make_frames`. However, here the `HTTP2_TRAILER_HEADERS` is actually a subclass of `HTTP2_HEADERS`, that is, it needs to be treated as header in that function too except some trivial changes specific to trailers. I just deliberately do this to reuse maximum code.